utils/yoke/src/yokeable.rs

```diff
@@ -0,0 +1,248 @@
+    /// data), that borrowed data is known to be valid
+    ///
+    /// Note that the `for<'b>` is also necessary, otherwise the following code would compile:
+    ///
+    /// ```rust,compile_fail
+    /// # use std::borrow::Cow;
+    /// # use yoke::Yokeable;
+    /// # use std::mem;
+    ///
+    /// // also safely implements Yokeable<'a>
+    /// struct Bar<'a> {
+    ///     num: u8,
+    ///     cow: Cow<'a, u8>,
+    /// }
+    ///
+    /// fn unsound<'a>(bar: &'a mut Bar<'static>) {
+    ///     bar.with_mut(move |bar| bar.cow = Cow::Borrowed(&bar.num));
+    /// }
+    ///
+    /// # unsafe impl<'a> Yokeable<'a> for Bar<'static> {
+    /// #     type Output = Bar<'a>;
+    /// #     fn transform(&'a self) -> &'a Bar<'a> {
+    /// #         self
+    /// #     }
+    /// #
+    /// #     unsafe fn make(from: Bar<'a>) -> Self {
+    /// #         let ret = mem::transmute_copy(&from);
+    /// #         mem::forget(from);
+    /// #         ret
+    /// #     }
+    /// #
+    /// #     fn with_mut<F>(&'a mut self, f: F)
+    /// #     where
+    /// #         F: 'static + FnOnce(&'a mut Self::Output),
+    /// #     {
+    /// #         unsafe { f(mem::transmute(self)) }
+    /// #     }
+    /// # }
+    /// ```
+    ///
+    /// which is unsound because `bar` could be moved later, and we do not want to be able to
+    /// self-insert references to it.
+    ///
+    /// Thus the only types of mutations allowed are ones that move around already-borrowed data, or
+    /// introduce new owned data:
+    ///
+    /// ```rust
+    /// # use std::borrow::Cow;
+    /// # use yoke::Yokeable;
+    /// struct Foo {
+    ///     str: String,
+    ///     cow: Cow<'static, str>,
+    /// }
+    ///
+    /// fn sound<'a>(foo: &'a mut Foo) {
+    ///     foo.cow.with_mut(move |cow| cow.to_mut().push('a'));
+    /// }
+    /// ```
+    fn with_mut<F>(&'a mut self, f: F)
+    where
+        // be VERY CAREFUL changing this signature, it is very nuanced (see above)
+        F: 'static + for<'b> FnOnce(&'b mut Self::Output);
+}
+
+unsafe impl<'a, T: 'static + ToOwned + ?Sized> Yokeable<'a> for Cow<'static, T>
+where
+    <T as ToOwned>::Owned: Sized,
+{
+    type Output = Cow<'a, T>;
+    fn transform(&'a self) -> &'a Cow<'a, T> {
+        // Doesn't need unsafe: `'a` is covariant so this lifetime cast is always safe
+        self
+    }
+
+    unsafe fn make(from: Cow<'a, T>) -> Self {
+        debug_assert!(mem::size_of::<Cow<'a, T>>() == mem::size_of::<Self>());
+        // i hate this
+        // unfortunately Rust doesn't think `mem::transmute` is possible since it's not sure the sizes
+        // are the same
+        let ret = mem::transmute_copy(&from);
+        mem::forget(from);
+        ret
+    }
+
+    fn with_mut<F>(&'a mut self, f: F)
+    where
+        F: 'static + for<'b> FnOnce(&'b mut Self::Output),
+    {
+        // Cast away the lifetime of Self
+        unsafe { f(mem::transmute::<&mut Self, &mut Self::Output>(self)) }
```

Comment:
I'm mildly neurotic about this kind of thing but I tend to put explicit lifetimes on lifetime-only transmutes just to be crystal-clear and avoid unbound lifetime hilarity.

Reply:
Fixed